### PR TITLE
[DO NOT MERGE] fix bug not restart event monitor correctly

### LINKED ISSUE #2309: swarm loses events if docker engine restarts!
Hi, All, 

This issue presents two parts:
- How to reproduce **Swarm loses control of docker's event**;
- How to locate the Swarm's code to identify code BUG;

First, Let me describe the issue:**Swarm will miss docker's events if we restart docker engine**.

The test environment has two parts: **One Swarm Binary**,**One Docker Engin**:

<h2>Steps to reproduce the issue:</h2>
1. Start Swarm Manager using `$GOPATH/bin/swarm --debug -l debug manage nodes://10.1.0.108:2376`;
2. Use `docker events` to listen Swarm Events;
3. Use `docker run` to create and start a container on  10.1.0.108, we can find `docker events` returns serveral events as wished;
4. On node 10.1.0.108, we run **service docker restart**;
5. Still on node 10.1.0.108, we `docker run` a container, `docker events` returns nothing.

**Version Details**:

```
~ $  docker version
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:44:17 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      swarm/1.2.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   HEAD
 Built:        <unknown>
 OS/Arch:      darwin/amd64
```

<h2>Locate the bug code</h2>

I found that Swarm uses **StartMonitorEvents** to listen on docker's event, if error happened restart listening via a recursive.

```

// StartMonitorEvents monitors events from the engine
func (e *Engine) StartMonitorEvents() {
    log.WithFields(log.Fields{"name": e.Name, "id": e.ID}).Debug("Start monitoring events")
    ec := make(chan error)
    e.eventsMonitor.Start(ec)

    go func() {
        if err := <-ec; err != nil {
            if !strings.Contains(err.Error(), "EOF") {
                // failing node reconnect should use back-off strategy
                <-e.refreshDelayer.Wait(e.getFailureCount())
            }
            e.StartMonitorEvents()
        }
        close(ec)
    }()
}
```

While I add more log in func StartMonitorEvents and func e.eventsMonitor.Start:

```
// StartMonitorEvents monitors events from the engine
func (e *Engine) StartMonitorEvents() {
    log.WithFields(log.Fields{"name": e.Name, "id": e.ID}).Debug("Start monitoring events")
    ec := make(chan error)
    e.eventsMonitor.Start(ec)

    go func() {
        if err := <-ec; err != nil {
            log.Infof("----------------------------------- %v", err)
            if !strings.Contains(err.Error(), "EOF") {
                // failing node reconnect should use back-off strategy
                <-e.refreshDelayer.Wait(e.getFailureCount())
            }
            e.StartMonitorEvents()
        }
        close(ec)
    }()
}
```

AND

```
// Start starts the EventsMonitor
func (em *EventsMonitor) Start(ec chan error) {
    em.stopChan = make(chan struct{})

    responseBody, err := em.cli.Events(context.Background(), types.EventsOptions{})
    if err != nil {
        fmt.Println(err.Error())
        ec <- err
        return
    }
        ...
}
```

Then save the files, compile a new binary and do the steps as Part 1. This time we will get log like this

```
INFO[0000] Registered Engine ubuntu at 10.1.0.108:2376
INFO[0013] ----------------------------------- unexpected EOF
DEBU[0013] Start monitoring events                       id=HRBI:EV6M:3NAC:CXZ3:BDS7:UOKL:PD6Y:ZCK4:V3PF:V7MR:CU2Y:3LGG name=ubuntu
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
DEBU[0056] Engine update succeeded                       id=HRBI:EV6M:3NAC:CXZ3:BDS7:UOKL:PD6Y:ZCK4:V3PF:V7MR:CU2Y:3LGG name=ubuntu
```

We found that StartMonitorEvents did not restart monitoring docker's event as we wish. Swarm just restarts monitoring once like the log showed us.

**Why does this start monitoring fail?**
For Swarm it almost takes no time to start one monitoring event, while for docker engine, it needs much more time to initialize itself. As a result, one start monitoring fails before docker engine start completely. **So, Swarm loses this engine's all events since then.**

I think it is a severe bug for Swarm. Hope one can try this, and then reproduce it. If this issue solved, Swarm will not lose some important events so that inconsistency state of container between Swarm and Docker Engine will disappear a lot.

related issue #2237 

cc @docker/core-swarm-maintainers ?

### PR DESCRIPTION
**Pull request by allencloud**
Fix a bug that Swarm missed events when a docker engine restarts.

Fixed #2309 

Signed-off-by: allencloud allen.sun@daocloud.io

### CODE REVIEW

**Comment by vieux** [comment]
ping @ehazlett

**Comment by ehazlett** [comment]
yay! :+1:

**Comment by allencloud** [comment]
I think this problem can cause a lot of inconsistency when docker restarts.
While I tried to solve this problem described in #2309, but my code in this PR may have some race condition in the travis CI test. And to be honest I am not quite familiar with how to solve the potential race in my code.  
I wish swarm can solve this issue as soon as possible. Another PR is totally welcomed as long as we solve this.
/cc @docker/core-swarm-maintainers

**Comment by yank1** [comment]
LDTM

File: cluster/engine.go [hunk 1]
```diff
@@ -190,25 +190,33 @@ func (e *Engine) Connect(config *tls.Config) error {
...
+			// Swarm will enter a loop to start eventsMonitor which may consume CPU a lot.
+			time.Sleep(1 * time.Second)
+
 			e.StartMonitorEvents()
 		}
-		close(ec)
 	}()
+
+	// start eventsMonitor as the first time.
+	e.RLock()
```
Comment (xianlubird) [inline cluster/engine.go]: Why lock engine ?

**Comment by allencloud** [comment]
PR updated. 
This PR has fixed the bug, and solved the potential race condition. I think it is really severe.
ping @vieux @ehazlett @abronan @dongluochen @nishanttotla @jimmyxian

File: cluster/event_monitor.go [hunk 2]
```diff
@@ -25,15 +27,36 @@ type decodingResult struct {
...
+
+			// EventMonitor ran into an error, start it again.
+			em.Start()
+		} else {
+			// if err is nil, it means stopChan is closed and EventMonitor should return as well.
+			return
+		}
+	}()
+
+	em.start(em.errChan)
```
Comment (dongluochen) [inline cluster/event_monitor.go]: There is no need to pass `errChan` to a method of `EventsMonitor`.

File: cluster/event_monitor.go [hunk 3]
```diff
@@ -25,15 +27,36 @@ type decodingResult struct {
...
+}
 
+// start starts to listen event stream of docker engine.
+func (em *EventsMonitor) start(ec chan error) {
```
Comment (dongluochen) [inline cluster/event_monitor.go]: AFAIK, if we use `start` and `Start`, they are same function with minor difference and one call the other. Here is not the case. I suggest choosing better name.

**Comment by dongluochen** [comment]
@allencloud Any progress on this? Please rebase.

**Comment by allencloud** [comment]
@dongluochen 
There is still some problem in this PR. I tried a lot in my local testing, about one in 20 will fail and continuous integration succeed on my local machine while fails in travis-ci.

**Comment by allencloud** [comment]
I updated this PR, while still **not the final one** with lots of comment in the PR.
It totally works on my local OS X system with `go test -v -race `go list ./... | grep -v /vendor/``, while failed in `continuous-integration/travis-ci/pr` in term of event. Could anyone help me to take a look?
@dongluochen @nishanttotla 

Below is my thoughts on this PR.

I think I should clarify some background of Swarm:

1-Two kinds of ways to start monitor swarm node's events:
* Engine first comes to swarm manager's consideration;
* When an engine's status become `Healthy` from `Unhealthy`; In additional, Only when the times of connection error reaches `e.opts.FailureRetry`, engine's status will be `Unhealthy`.

2-two ways to stop eventMonitor of engine
* Swarm manager regards engine is disconnected
* Swarm manager regards it as unhealthy

In the implementation, there will be two import goroutines:
The first one is:
```
go func() {
		for {
			select {
			case err := <-em.errChan:
				if err != nil {
					// Swarm needs to wait an interval to start eventsMonitor.
					// If the Engine is lost forever, before Swarm judges it as disconnected，
					// Swarm will enter a loop to start eventsMonitor which may consume CPU a lot.
					time.Sleep(1 * time.Second)

					// EventMonitor ran into an error, start it again.
					go em.startEventsMonitoring()

				} else {
					// if err is nil, it means stopChan is closed and EventMonitor should return as well.
					return
				}
			case <-em.stopChan:
				// if err is nil, it means stopChan is closed and EventMonitor should return as well.
				return
			}
		}
	}()
```

The second one is : 
```
go func() {
		defer responseBody.Close()
		dec := json.NewDecoder(responseBody)

		for {
			select {
			case <-em.stopChan:
				return
			default:
				var result decodingResult
				result.err = dec.Decode(&result.msg)
				if result.err != nil {
					em.errChan <- result.err
					return
				}
				if err := em.handler(result.msg); err != nil {
					em.errChan <- err
					return
				}
			}
		}

```

One import thing is that, when stop eventMonitor, every goroutine should exit. There is two goroutines above, so in the `func (em *EventsMonitor) Stop()`, stopChan should has a length of 2 for each goroutine. This will guarantee that, Once stop, all exit. There when startMonitor again, everything like before as setup all things.

**Comment by nishanttotla** [comment]
@allencloud would you like to update this PR now? Hopefully we can then get it into the next release.

**Comment by nishanttotla** [comment]
@allencloud can you please rebase this PR?

**Comment by nishanttotla** [comment]
@allencloud do you think we should move this PR to the next milestone, or would you prefer to get this into 1.2.6?

cc @dongluochen is it an issue if this doesn't make 1.2.6?

**Comment by realityone** [comment]
@nishanttotla 
The event monitor can't restart correctly as expect because of a common mistake in golang at line `cluster/engine.go:208`, this PR is refactoring these code to be more elegant, but a work around solution is just fix this.

As this bug has been very long time in swarm, consider review this PR https://github.com/docker/swarm/pull/2547 for 1.2.6, this is just a work around solution.

**Comment by dongluochen** [comment]
@allencloud Thanks for reporting the problem. This PR was outdated with recent changes. #2583 should fix the problem.

**Pull request closed without merging**